numpy/core/function_base.py

```diff
@@ -123,25 +123,47 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     if num < 0:
         raise ValueError("Number of samples, %s, must be non-negative." % num)
     div = (num - 1) if endpoint else num
+    # Make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start)
+    stop  = asanyarray(stop)
+
+    # check if dealing with datetime64 or timedelta64
+    is_datelike = start.dtype.kind in 'Mm'
+    if is_datelike:
+        # For datetime/timedelta, dtype must be specified
+        # start and stop will first be cast to this dtype.
+        # Warning: dtype must have high enough precision so
+        # that it is at all possible to draw num equispaced
+        # samples.
+        assert dtype is not None, "linspace with {} objects cannot be called "\
+            "with dtype=None".format(start.dtype)
+        start = start.astype(dtype)
+        stop = stop.astype(dtype)
+        # y cannot be cast to datetime64/timedelta64, but left
+        # as int64, since later on we will do y * delta, and
+        # <m8[s] * <m8[s] is not allowed.
+        delta = stop - start
+        y = _nx.arange(0, num).reshape((-1,) + (1,) * ndim(delta))
 
-    # Convert float/complex array scalars to float, gh-3504
-    # and make sure one can use variables that have an __array_interface__, gh-6634
-    start = asanyarray(start) * 1.0
-    stop  = asanyarray(stop)  * 1.0
+    else:
+        # Convert float/complex array scalars to float, gh-3504
+        start = start * 1.0
+        stop  = stop  * 1.0
+        dt = result_type(start, stop, float(num))
+        delta = stop - start
+        y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
 
-    dt = result_type(start, stop, float(num))
-    if dtype is None:
-        dtype = dt
+        if dtype is None:
+            dtype = dt
 
-    delta = stop - start
```

Comment:
If would be nice if we could do this without any special casing. Does the following work?
```
start = asanyarray(start)
stop  = asanyarray(stop)
dt = result_type(start, stop, float(num))
if dtype is None:
    dtype = dt
delta = stop.astype(dt) - start.astype(dt)
```

Reply:
Unfortunately not. `result_type(start, stop, float(num))` raises `TypeError` if start/stop are `datetime64` since type promotion between `datetime64` and `float` are not allowed.

Reply:
How about
```
start = asanyarray(start)
stop  = asanyarray(stop)
delta = stop - start
dt = result_type(delta, float(num))
delta = delta.astype(dt)
if dtype is None:
    dtype = dt
```

Reply:
This *would* have worked if type promotion between `timedelta64` and `float` resulted in `float`, but it doesn't:
```python
>>> np.result_type(np.timedelta64(1, 'D'), float(1))                                   
dtype('<m8[D]')
```
Is this type promotion the intended behaviour? I do not know numpy's internals well enough to judge this.

Reply:
@seberg : Could you please comment on this if you have time?

Reply:
How about `ndarray.view`? Since `timedelta64` is just `int64` with units attached (essentially), we can get the desired `float` by, e.g., `np.asarray(np.timedelta64(1, 'D')).view(np.int64)`. And then dividing by the appropriate amount if we want to normalise to e.g. seconds.

Reply:
I think the `result_type` not returning float is definitely right. However, one could argue that it should just fail in the generic case (and only work as promotion in certain ufuncs). Hameers suggestion to use the knowledge that it is int64 may work. Although have to be careful with `NaT`.

Reply:
@hameerabbasi @seberg : Thank you for your comments. I have tried the `ndarray.view` approach and it works. Adopting this led to reducing the number of `if/else` blocks and improved readability.